pkg/services/ngalert/eval/eval.go

```diff
@@ -433,17 +438,10 @@ func queryDataResponseToExecutionResults(c models.Condition, execResp *backend.Q
 	}
 
 	result := ExecutionResults{Results: make(map[string]data.Frames)}
-	for refID, res := range execResp.Responses {
-		if res.Error != nil {
-			if result.Errors == nil {
-				result.Errors = make(map[string]error)
-			}
-			result.Errors[refID] = res.Error
-			if refID == c.Condition {
-				result.Error = res.Error
-			}
-		}
 
+	result.Error = FindConditionError(execResp, c.Condition)
```

Comment:
Extracting this to another function means we now loop over the responses more than once needlessly. Not ideal, but there are usually a low number of frames per response in practice.

This function was already so complex it needed a cyclomatic complexity suppression. Doing this means we can drop that linter suppression. Plus, IMO it's worth it having both rule types use exactly the same code path, there is less chance for discrepancy.

Reply:
What's the upper bound look like on the added overhead here? I wonder if breaking out the logic at the Response level (so within the loop) might remove the need for a second loop and get us the same benefit.